### ISSUE TITLE
Hold løbende styr på hvem der er online på IRC

### ISSUE THREAD

**Issue by nqpz** [Athas/EggsML#39]
concieggs bruger lige nu eggspi-kommandoen `recentlyActive` til at få en liste af nyligt aktive brugere baseret på hvem der har skrevet noget i de sidste N linjer.  Det bruges til bl.a. `flaske`- og `fluen`-kommandoen.

Det virker fint, bortset fra at nogle af de aktive brugere kan være logget af siden de har skrevet deres besked, hvilket kan være suboptimalt.  Det er fx noget rod at krone en som slet ikke er online!

Problemet er at det tager lang tid at tjekke om nogen er online: concieggs skal sende en WHOIS til freenode for hver bruger, så det kan godt blive 5-10 whoises, og det er langsomt.

Det kan løses ved at det registreres når brugere logger på og af, så det gemmes i concieggs' database hvem der er online.  Så er det trivielt at tjekke om nogen er på IRC eller ej.  Der skal ikke gemmes hvornår folk er logget på og af, kun hvorvidt de er online eller offline.

**Comment by Athas** [comment]
"Niels G. W. Serup" <notifications@github.com> writes:

> Det kan løses ved at det registreres når brugere logger på og af, så
> det gemmes i concieggs' database hvem der er online.  Så er det
> trivielt at tjekke om nogen er på IRC eller ej.

Ligesom en voksion-IRC-klient!

-- 
\  Troels
/\ Henriksen

**Comment by Eckankar** [comment]
Det ville også være fedt hvis den trackede det pr. kanal!

**Comment by sshine** [comment]
Ja, eller at `flaske` var specifik for de senest aktive brugere på en given kanal.

**Comment by svip** [comment]
Hvorfor er det så svært at implementere?  Er vi bare dovne?

**Comment by sshine** [comment]
Jeg har forsøgt at markere problemet som vigtigt.

Særligt har jeg maskeret det som et krav til opfyldelse af GDPR, fordi det er den nærmeste deadline vi har.

**Comment by sshine** [comment]
Der er nu [`hooks/channel_join/05register_join`](https://github.com/Athas/EggsML/blob/master/concieggs/hooks/channel_join/05register_join) og [`hooks/channel_part/05register_part`](https://github.com/Athas/EggsML/blob/master/concieggs/hooks/channel_part/05register_part).

- [ ] Gør krogene atomiske.
- [ ] Gem også hvornår nogen er kommet ind.
- [ ] Registrér når det er concieggs der joiner.
- [ ] Håndtér 353/366, så concieggs kan genopfriske listerne uden at alle forlader og kommer igen:

        353     RPL_NAMREPLY
                        "<channel> :[[@|+]<nick> [[@|+]<nick> [...]]]"
        366     RPL_ENDOFNAMES
                        "<channel> :End of /NAMES list"

                - To reply to a NAMES message, a reply pair consisting
                  of RPL_NAMREPLY and RPL_ENDOFNAMES is sent by the
                  server back to the client.  If there is no channel
                  found as in the query, then only RPL_ENDOFNAMES is
                  returned.  The exception to this is when a NAMES
                  message is sent with no parameters and all visible
                  channels and contents are sent back in a series of
                  RPL_NAMEREPLY messages with a RPL_ENDOFNAMES to mark
                  the end.

**Comment by MikkelStorgaard** [comment]
On Wed, May 2, 2018 at 9:53 AM, Simon Shine <notifications@github.com>
wrote:

> Der er nu hooks/channel_join/05register_join
> <https://github.com/Athas/EggsML/blob/master/concieggs/hooks/channel_join/05register_join>
> og hooks/channel_part/05register_part
> <https://github.com/Athas/EggsML/blob/master/concieggs/hooks/channel_part/05register_part>
> .
>
>    -
>
>    Gør krogene atomiske.
>    -
>
>    Gem også hvornår nogen er kommet ind.
>    -
>
>    Registrér når det er concieggs der joiner.
>    -
>
>    Håndtér 353/366, så concieggs kan genopfriske listerne uden at alle
>    forlader og kommer igen:
>
>      353     RPL_NAMREPLY
>                      "<channel> :[[@|+]<nick> [[@|+]<nick> [...]]]"
>      366     RPL_ENDOFNAMES
>                      "<channel> :End of /NAMES list"
>
>              - To reply to a NAMES message, a reply pair consisting
>                of RPL_NAMREPLY and RPL_ENDOFNAMES is sent by the
>                server back to the client.  If there is no channel
>                found as in the query, then only RPL_ENDOFNAMES is
>                returned.  The exception to this is when a NAMES
>                message is sent with no parameters and all visible
>                channels and contents are sent back in a series of
>                RPL_NAMEREPLY messages with a RPL_ENDOFNAMES to mark
>                the end.
>
>
> Hvordan kan vi afgøre, om en given bruger ikke siger så meget men bare
læser med på kanalen, eller reelt set ikke bruger kanalen, og bare er der?

**Comment by sshine** [comment]
@MikkelStorgaard: For at citere den oprindelige problembeskrivelse:

> Der skal *[...]* kun [gemmes] hvorvidt [folk] er online eller offline.

Men du kan afgøre om nogen for nyligt har sagt noget vha. `recentlyActive` og om de er online ved at læse `db/store/on_channel/#diku`. (Der er endnu ikke nogen eggspi-kommandoer.)

**Comment by nqpz** [comment]
On Wed, 02 May 2018 10:11:04 +0000, Simon Shine wrote:
> @MikkelStorgaard: For at citere den oprindelige problembeskrivelse:
> 
> > Der skal *[...]* kun [gemmes] hvorvidt [folk] er online eller offline.
> 
> Men du kan afgøre om nogen for nyligt har sagt noget vha. `recentlyActive` og om de er online ved at læse `db/store/on_channel/#diku`. (Der er endnu ikke nogen eggspi-kommandoer.)

Og det vil helt klart give mening at lave en samlet eggspi-kommando `isActive` der laver nogle gode antagelser og bruger data fra begge kilder.

**Comment by sshine** [comment]
Eller gemmer hvornår man sidst har været aktiv i `db/store/users/<bruger>/lastactive` ved siden af filen `lastmsg`. Man kan egentlig bruge tidsstemplet på `lastmsg`, da filen ikke ændres af [`s//`-krogen](https://github.com/Athas/EggsML/blob/master/concieggs/hooks/channel_message/40regex) med mindre det er personen selv der aktiverer den, men det er måske ikke en garanti vi ønsker at fastkode.

**Comment by sshine** [comment]
Problemet er mestendels løst vha.

1. Krogene [`hooks/channel_join/05register_join`](https://github.com/Athas/EggsML/blob/master/concieggs/hooks/channel_join/05register_join) og [`hooks/channel_part/05register_part`](https://github.com/Athas/EggsML/blob/master/concieggs/hooks/channel_part/05register_part).

2. Den er nu gjort bevidst om når den selv joiner/parter: f66f7334983babeb95a0f3a1cc652a7a254c89c5. [Når den gør det](https://github.com/Athas/EggsML/commit/aaaa17254b31c098e6b1ace12abac72d61a0bd31#diff-532d0a02999ce22c38b47a665e2cf926), kalder den [`eggspi/getNamesOnChannel [#kanal]`](https://github.com/Athas/EggsML/blob/master/concieggs/eggspi/getNamesOnChannel), som scanner `$CONCIEGGS_IRC_DIR/out` baglæns efter sidste gang der er kaldt `/NAMES` på `#kanal` og printer navnene. Det er fordi `/NAMES` bliver kørt automatisk når man joiner.

- [ ] Den registrerer stadig ikke når folk quitter.
- [ ] Den fjerner ikke al viden om hvem som er på kanalen, når den selv forlader en kanal.

**Comment by svip** [comment]
Så skal det være muligt at spørge concieggs:

```concieggs: er der nogle frække fyrer på linje?```

**Comment by nqpz** [comment]
Og
```
<preben> concieggs: giv mig en ven
<concieggs> preben: sigurd er din nye ven!
```
